pilot/pkg/serviceregistry/kube/controller/serviceimportcache.go

```diff
@@ -111,7 +111,7 @@ func (ic *serviceImportCacheImpl) onServiceEvent(svc *model.Service, event model
 	// service has a ServiceImport in this cluster.
 	vips := ic.getClusterSetIPs(namespacedName)
 
-	if event == model.EventDelete || len(vips) == 0 {
+	if len(vips) == 0 || (event == model.EventDelete && ic.opts.MeshServiceController.GetService(kube.ServiceHostname(namespacedName.Name, namespacedName.Namespace, ic.opts.DomainSuffix)) == nil) {
```

Comment:
do we event need to look for `model.EventDelete`? seems like we can/should just reconcile to the current state

Reply:
> seems like we can/should just reconcile to the current state

Yes but I think the idea is to clean up mcs
We [create mcs only if the real services exist](https://github.com/istio/istio/blob/1.13.1/pilot/pkg/serviceregistry/kube/controller/serviceimportcache.go#L164-L172), so it makes sense to delete it when all real services are gone